Netkan/Services/CachingHttpService.cs

```diff
@@ -0,0 +1,141 @@
+
+            _curl = new CurlEasy { UserAgent = CKAN.Net.UserAgentString };
+
+            var caBundle = ResolveCurlCaBundle();
+            if (caBundle != null)
+            {
+                _curl.CaInfo = caBundle;
+            }
+        }
+
+        public string DownloadPackage(Uri url, string identifier)
+        {
+            EnsureNotDisposed();
+
+            return _cache.GetCachedFilename(url) ?? 
+                _cache.Store(url, CKAN.Net.Download(url), string.Format("netkan-{0}", identifier), move: true);
+        }
+
+        public string DownloadText(Uri url)
+        {
+            EnsureNotDisposed();
+
+            Log.DebugFormat("About to download {0}", url);
+
+            var content = string.Empty;
+
+            _curl.Url = url.ToString();
+            _curl.WriteData = null;
+            _curl.WriteFunction = delegate(byte[] buf, int size, int nmemb, object extraData)
+            {
+                content += Encoding.UTF8.GetString(buf);
+                return size * nmemb;
+            };
+
+            var result = _curl.Perform();
+
+            if (result != CurlCode.Ok)
+            {
+                throw new Exception("Curl download failed with error " + result);
+            }
+
+            Log.DebugFormat("Download from {0}:\n\n{1}", url, content);
+
+            return content;
+        }
+
+        /// <summary>
+        /// Resolves the location of the cURL CA bundle file to use.
+        /// </summary>
+        /// <returns>The absolute file path to the bundle file or null if none is found.</returns>
+        private static string ResolveCurlCaBundle()
+        {
+            const string caBundleFileName = "curl-ca-bundle.crt";
+            const string ckanSubDirectoryName = "CKAN";
+
+            var bundle = new[]
+            {
+                // Working Directory
+                Environment.CurrentDirectory,
+
+                // Executable Directory
+                Path.GetDirectoryName(Assembly.GetExecutingAssembly().Location),
+
+                // %LOCALAPPDATA%/CKAN
+                Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData), ckanSubDirectoryName),
+
+                // %APPDATA%/CKAN
+                Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.ApplicationData), ckanSubDirectoryName),
+
+                // %PROGRAMDATA%/CKAN
+                Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.CommonApplicationData), ckanSubDirectoryName),
+            }
+            .Select(i => Path.Combine(i, caBundleFileName))
+            .FirstOrDefault(File.Exists);
+
+            Log.InfoFormat("Using curl-ca bundle: {0}",bundle ?? "(none)");
+
+            return bundle;
+        }
+    }
+
+    internal sealed partial class CachingHttpService
+    {
+        private bool _isDisposed;
+
+        ~CachingHttpService()
+        {
+            Dispose(disposing: false);
```

Comment:
Pretty much spot on. The destructor will be called if the object is about to be garbage collected without someone previously calling `Dispose()` (usually via a `using` statement). This is pretty much a canonical implementation of the `IDisposable` [pattern](https://msdn.microsoft.com/en-us/library/ms244737.aspx). The missing bit is the clean-up of native resources, since we don't actually have any native resources (The cURL lib presumably does, but we're not responsible for any of that).

Managed resources are only cleaned up when `disposing` is `true`, since if we're being called by the finalizer, you can't assume *any* managed resource still exists.

I am now just seeing that I'm missing the call to `GC.SuppressFinalize(object)`. Not strictly necessary, but a small micro-optimization that prevents the finalizer from bothering to call the destructor if the object was already disposed. I'll add it in a bit.